### DISCUSSION TITLE
Calculation behind spectrum plots

### DISCUSSION THREAD

**Discussion by AishwaryaMukund** [audacity/audacity#908] [Q&A] [answered]
Hello, 

I am trying to plot spectrum using Python' Librosa library, but it doesn't match the audacity plot. I tried to dig through code to check for calculations and they are all in Lisp (which I am not familiar with). Does anyone have any resources on how to get that  spectrum plot ? Resources could be just Math calculations or Python code.

**Comment by Paul-Licameli** [comment, marked as answer]
Are you referring to Lisp code in Audacity?  There is some, but it does not calculate the spectrogram.

Are you referring to spectrogram views of wave tracks, or the graphs that you see with the Plot Spectrum... analyzer?

There is no single spectrogram of a sound.  The plot you get depends on many choices, such as the window size, the windowing function, zero padding, whether you choose the Reassignment algorithm, and so on.

Here is more information.  https://manual.audacityteam.org/man/spectrogram_view.html

**Reply by AishwaryaMukund** [reply]
I was referring to the plot spectrum analyzer. I am trying to plot this in python but not able to produce this graph. So was wanted to know what calculations did they use to plot this at default settings of audacity. Mainly, how is the `db` values calculated?
![image](https://user-images.githubusercontent.com/35746763/118880029-0063d280-b8b7-11eb-9493-62baed670cc5.png)

**Reply by JamesCrook** [reply]
Try source code from here:
https://github.com/audacity/audacity/blob/master/src/FreqWindow.cpp#L780
and from here:
https://github.com/audacity/audacity/blob/master/src/SpectrumAnalyst.cpp#L93

So C++, not Lisp/Nyquist.

**Reply by boim23Raptor** [reply]
Hallo @JamesCrook 

I was following the 
https://github.com/audacity/audacity/blob/master/src/SpectrumAnalyst.cpp#L93
so I use the Hann Window 
if (i < (FFT_Len - 1)) {
            window = 0.5 - (0.5 * cos(2 * PI * i / FFT_Len)); //sin(PI * i / FFT_Len) * sin(PI * i / FFT_Len);
        }
        else {
            window = 1;
        }

do 2 x FFT every FFT length,  e.g. we have data length 5FFT, then every half fft length i do fft, so we will have 5 x2 FFT samples per carrier., then i do scaling with wss = sum(window); wss =4/(wss*wss)
basically the sum of the 10 freq. domain samples is averaged then multiply by wss, then the the 10 * log10(value)

But I have discrepancy  on the figure, on the closed bands

is there anything else, that i miss.

Kind regards

I. Budiarjo

[Spectra from Audacity.pdf](https://github.com/audacity/audacity/files/13437565/Spectra.from.Audacity.pdf)
[Spectra from Audacity.pdf](https://github.com/audacity/audacity/files/13437562/Spectra.from.Audacity.pdf)
